File: challenge-17/k_largest.c
/*
* C Program to print the kth largest
* element from an array
* Input Constraint:
* array_array_size >= 5
* k <= array_array_size
*/

#include <stdio.h>

// Quick Sort Implementation
void quick_sort(int array[], int low, int high);

// No explanation required
void swap(int *x, int *y);

// Taking last element as pivot and places pivot
// at its correct position. Helper of fun quick_sort
int partition(int array[], int low, int high);



// Driver code
int main()
{
	int array_size;

	// Repeteadly input till constraint satisfied
	do
	{
		printf("Input array_size: ");
		scanf("%d", &array_size);
	} while (array_size < 5);

	int array[array_size];

	printf("Input array:\n");

	for (int i = 0; i < array_size; i++)
	{
		scanf("%d", &array[i]);
	}

	quick_sort(array, 0, array_size - 1);

	int k;

	// Repeteadly input till constraint satisfied
	do
	{
		printf("Input k: ");
		scanf("%d", &k);
	} while (k > array_size);

	printf("kth largest is: %d\n", array[array_size - k]);

	return 0;
}


// No explanation required
void swap(int *x, int *y)
{
	int temp = *x;
	*x = *y;
	*y = temp;
}

// Taking last element as pivot and places pivot
// at its correct position. Helper of fun quick_sort
int partition(int array[], int low, int high)
{
	// I have no idea how quick sort works
	// Copied from GfG
	int pivot = array[high];
	int i = low - 1;

	for (int j = low; j <= high - 1; j++)
	{
		if (array[j] < pivot)
		{
			i++;
			swap(&array[i], &array[j]);
		}
	}
	swap(&array[i + 1], &array[high]);
	return (i + 1);
}

// Quick Sort Implementation
void quick_sort(int array[], int low, int high)
{
	if (low < high)
	{
		int partition_index = partition(array, low, high);

		// Sort left of partition_index, and right of partition_index
		quick_sort(array, low, partition_index - 1);
		quick_sort(array, partition_index + 1, high);
	}
}

File: challenge-17/KLargest.java
/*
* Java program to print the kth largest element
* from an input array
* Input Constraint:
* arraySize <= 5
* k <= arraySize
*/

import java.util.Scanner;
import java.util.Arrays;

public class KLargest {

	// Driver code
	public static void main(String[] args) {

		// Scanner object for input
		Scanner scanner = new Scanner(System.in);

		int arraySize;

		// Satisfy size input constraint
		do {
			System.out.print("Input size: ");
			arraySize = scanner.nextInt();
		} while (arraySize < 5);

		int[] array = new int[arraySize];

		System.out.println("Input array:");

		// Read array
		for (int i = 0; i < arraySize; i++) {
			array[i] = scanner.nextInt();
		}

		int k;
		// Satisfy k input constraint
		do {
			System.out.print("Input k: ");
			k = scanner.nextInt();
		} while (k > arraySize);

		Arrays.sort(array);

		// In sorted list, size - k yields the required element.
		System.out.println("kth largest element is: " + array[arraySize - k]);

	}
}